scheduling/storage_providers/memory.go

```diff
@@ -2,48 +2,187 @@ package scheduling
 }
 
-func (m *MemoryStorageProvider) Set(key string, value string) error {
+func (m *MemoryStorageProvider) Set(key string, value string, lock LockObject) error {
 	if key == "" {
 		return fmt.Errorf("key is empty")
 	}
 	if value == "" {
 		return fmt.Errorf("value is empty for key %s", key)
 	}
-	m.storage[key] = value
+
+	lockInfo, ok := m.lockedItems.Load(key)
+	if !ok {
+		return fmt.Errorf("key is not locked")
+	}
+
+	currentLock := lockInfo.(LockInformation)
+	if currentLock.ID != lock.ID {
+		return fmt.Errorf("key %s is locked by another id: locked by: %s, unlock by: %s", key, currentLock.ID, lock.ID)
+	}
+
+	m.storage.Store(key, value)
 	return nil
 }
 
-func (m *MemoryStorageProvider) Get(key string, prefix bool) ([]string, error) {
+func (m *MemoryStorageProvider) Get(key string, prefix bool) (map[string]string, error) {
 	if key == "" {
 		return nil, fmt.Errorf("key is empty")
 	}
+
+	result := make(map[string]string)
+
 	if !prefix {
-		value, ok := m.storage[key]
+		value, ok := m.storage.Load(key)
 		if !ok {
 			return nil, &KeyNotFoundError{Key: key}
 		}
-		return []string{value}, nil
+		result[key] = value.(string)
+		return result, nil
 	}
 
-	var result []string
-	for k, v := range m.storage {
-		if strings.HasPrefix(k, key) {
-			result = append(result, v)
+	// loops through the keys in sync map
+	// and finds the keys that have the prefix
+	m.storage.Range(func(k, v interface{}) bool {
+		mapKey := k.(string)
+		if strings.HasPrefix(mapKey, key) {
+			result[mapKey] = v.(string)
 		}
-	}
+		return true
+	})
+
 	if len(result) == 0 {
 		return nil, &KeyNotFoundError{Key: key}
 	}
 	return result, nil
 }
+
+func (m *MemoryStorageProvider) ListKeys(prefix string) ([]string, error) {
+	var result []string
+	// loops through the keys in sync map
+	// and finds the keys that have the prefix
+	m.storage.Range(func(k, v interface{}) bool {
+		mapKey := k.(string)
+		if strings.HasPrefix(mapKey, prefix) {
+			result = append(result, mapKey)
+		}
+		return true
+	})
+	sort.Strings(result)
+
+	return result, nil
+}
+
+func (m *MemoryStorageProvider) Lock(key string) (LockObject, error) {
+	if key == "" {
+		return LockObject{}, fmt.Errorf("key is empty")
+	}
+
+	id := uuid.New().String()
+
+	lockMutex := &sync.Mutex{}
+	lockMutex.Lock()
+	defer lockMutex.Unlock()
+
+	if lockInfo, ok := m.lockedItems.Load(key); ok {
+		keyLock := lockInfo.(LockInformation)
+		if time.Since(keyLock.Date) < ValidTimePeriod {
+			return LockObject{}, fmt.Errorf("key is already locked by: %s", keyLock.ID)
+		}
+	}
+
+	lockChannel := make(chan error)
+	go m.updateLockTime(id, key, lockChannel)
+	lockObject := LockObject{ID: id, Channel: &lockChannel}
+
+	lock := LockInformation{
+		ID:   id,
+		Key:  key,
+		Date: time.Now(),
+		Lock: lockObject,
+	}
+	m.lockedItems.Store(key, lock)
+
+	return lockObject, nil
+}
+
+func (m *MemoryStorageProvider) Unlock(key string, lock LockObject) error {
+	lockMutex := &sync.Mutex{}
```

Comment:
wait. I don't think this is right.

What does the lockMutex do here?

You create a new lockMutex on every call to this function. So... each call to this function has a new mutex? Doesn't there need to be shared mutex for this all to work?